presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -831,9 +880,92 @@ public long getTimeSpentWaitingForOutputInMillis()
         }
     }
 
+    private static class DiskPageOutputSupplier
+            implements OutputSupplier<PrestoSparkDiskPage>
+    {
+        private static final MutablePartitionId DEFAULT_PARTITION = new MutablePartitionId();
+
+        private final PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer;
+        private final TempStorage tempStorage;
+        private final TempDataOperationContext tempDataOperationContext;
+        private TempDataSink tempDataSink;
+
+        private long timeSpentWaitingForOutputInMillis;
+
+        private DiskPageOutputSupplier(PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer,
+                TempStorage tempStorage,
+                TempDataOperationContext tempDataOperationContext)
+        {
+            this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
+            this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+            this.tempDataOperationContext = requireNonNull(tempDataOperationContext, "tempDataOperationContext is null");
+
+            try {
+                tempStorage.initialize(false);
+            }
+            catch (IOException e) {
+                throw new RuntimeException("Unable to initialize tempStorage: ", e);
+            }
+        }
+
+        @Override
+        public Tuple2<MutablePartitionId, PrestoSparkDiskPage> getNext()
```

Comment:
The implementation is going to block for a very long time before it returns anything. Is Spark going to be okay with this?